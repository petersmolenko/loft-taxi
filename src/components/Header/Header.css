.Header {
    padding: .2rem 1.5rem;
    background: white;
    /* min-width: 100vw; */
    display: flex;
    box-shadow: 0 .2rem .5rem rgba(0, 0, 0, .2);
}

.Header__logo {
    display: flex;
    flex-grow: 1;
    justify-content: flex-start;
    align-items: center;
    font-size: 1rem;
}

.Header__logo-img {
    width: 156px;
}

.Header__controls {
    display: flex;
    justify-content: center;
    align-items: center;
}

.Header__btn {
    background: inherit;
    font-family: 'Roboto', sans-serif;
    line-height: 1.75;
    font-size: 0.875rem;
    font-weight: 500;
    border: none;
    outline: none;
    padding: .375rem .7rem;
    border-radius: .3rem;
    color: rgba(0, 0, 0, 0.87);
    letter-spacing: 0.02857em;
}

.Header__btn:hover {
    cursor: pointer;
    background: rgba(0, 0, 0, 0.08);
}