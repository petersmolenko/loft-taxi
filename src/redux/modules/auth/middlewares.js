import { loggedIn, loggedInSuccess, loggedInFailure, signUp } from "./actions";

export const loginMiddleware = (store) => (next) => (action) => {
    const result = next(action);
    if (action.type === loggedIn.toString()) {

        fetch("https://loft-taxi.glitch.me/auth", {
            method: "POST",
            headers: {
                "Content-Type": "application/json;charset=utf-8",
            },
            body: JSON.stringify(action.payload),
        })
            .then((resp) => resp.json())
            .then((auth) => {
                if (auth.success === true) {
                    store.dispatch(loggedInSuccess(auth.token));
                } else {
                    store.dispatch(loggedInFailure(auth.error));
                }
            })
            .catch((error) => {
                store.dispatch(loggedInFailure("Error!!"));
            });
    }

    return result;
};

export const signupMiddleware = (store) => (next) => (action) => {
    const result = next(action);
    if (action.type === signUp.toString()) {
        console.log(action.payload);
        fetch("https://loft-taxi.glitch.me/register", {
            method: "POST",
            headers: {
                "Content-Type": "application/json;charset=utf-8",
            },
            body: JSON.stringify(action.payload),
        })
            .then((resp) => resp.json())
            .then((auth) => {
                if (auth.success === true) {
                    store.dispatch(loggedInSuccess(auth.token));
                } else {
                    store.dispatch(loggedInFailure(auth.error));
                }
            })
            .catch((error) => {
                store.dispatch(loggedInFailure("Error!!"));
            });
    }

    return result;
};
