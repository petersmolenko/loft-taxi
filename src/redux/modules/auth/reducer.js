import {
    loggedOut,
    loggedIn,
    loggedInSuccess,
    loggedInFailure,
    signUp,
    clearErrors
} from "./actions";

export default (state = {}, action) => {
    switch (action.type) {
        case clearErrors.toString():
            return { ...state, error: null };
        case loggedOut.toString():
            return { ...state, isLoggedIn: false };
        case loggedIn.toString():
            return { ...state, isLoaded: true };
        case signUp.toString():
            return { ...state, isLoaded: true };
        case loggedInSuccess.toString():
            return {
                ...state,
                isLoggedIn: true,
                token: action.payload,
                isLoaded: false,
                error: null,
            };
        case loggedInFailure.toString():
            return { ...state, isLoaded: false, error: action.payload };
        default:
            return state;
    }
};
