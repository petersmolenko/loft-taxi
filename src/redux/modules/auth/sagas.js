import { call, put, takeLatest } from "redux-saga/effects";
import { loggedIn, loggedInSuccess, loggedInFailure, signUp } from "./actions";
import { fetchProfile } from "../profile/actions";
import {fetchAddressList} from '../routes/actions';
import {auth as apiAuth, register as apiRegister} from '../../../shared/api'

export function* authorizationSaga(action) {
    try {
        const auth = yield call(apiAuth, action.payload);

        if (auth.success === true) {
            yield put(loggedInSuccess(auth.token));
            yield put(fetchProfile());
            yield put(fetchAddressList());
        } else {
            yield put(loggedInFailure(auth.error));
        }
    } catch (e) {
        yield put(loggedInFailure(e.message));
    }
}

export function* watchAuthorizationSaga() {
    yield takeLatest(loggedIn.toString(), authorizationSaga);
}

export function* registrationSaga(action) {
    try {
        const auth = yield call(apiRegister, action.payload);
        
        if (auth.success === true) {
            yield put(loggedInSuccess(auth.token));
            yield put(fetchAddressList());
        } else {
            yield put(loggedInFailure(auth.error));
        }
    } catch (e) {
        yield put(loggedInFailure(e.message));
    }
}

export function* watchRegistrationSaga() {
    yield takeLatest(signUp.toString(), registrationSaga);
}
