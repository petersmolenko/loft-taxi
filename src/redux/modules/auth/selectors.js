const isLoggedIn = state => state.auth.isLoggedIn
const getToken = state => state.auth.token

export {isLoggedIn, getToken}