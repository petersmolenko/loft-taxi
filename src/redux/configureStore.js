import { createStore, applyMiddleware, combineReducers } from 'redux';
import auth, {loginMiddleware, signupMiddleware} from './modules/auth'
import profile, {fetchProfileMiddleware} from './modules/profile'

const createStoreWithMiddleware = applyMiddleware( loginMiddleware, signupMiddleware, fetchProfileMiddleware)(createStore); 

const reducer = combineReducers({
  auth, profile
});

const configureStore = (initialState) => createStoreWithMiddleware(reducer, initialState);
export default configureStore;