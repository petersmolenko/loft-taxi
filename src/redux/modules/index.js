import { combineReducers } from "redux";
import { fork } from "redux-saga/effects";
import auth from "./auth";
import profile from "./profile";
import routes from "./routes";
import { watchFetchProfileSaga, watchPutProfileSaga } from "./profile";
import {
    watchRouteSaga,
    watchAddressListSaga,
} from "./routes";
import {
    watchAuthorizationSaga,
    watchRegistrationSaga,
} from "./auth";

export default combineReducers({ auth, profile, routes });

export function* rootSaga() {
    yield fork(watchFetchProfileSaga);
    yield fork(watchPutProfileSaga);
    yield fork(watchRouteSaga);
    yield fork(watchAddressListSaga);
    yield fork(watchAuthorizationSaga);
    yield fork(watchRegistrationSaga);
}
