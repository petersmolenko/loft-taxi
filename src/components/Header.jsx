import React from "react";
import { Grid, Paper, Button } from "@material-ui/core";
import logo from "../assets/logo.svg";
import { Link } from "react-router-dom";
import { loggedOut } from "../store/modules/auth";
import { clearRoute } from "../store/modules/routes";
import { useDispatch } from "react-redux";

let Header = () => {
    const dispatch = useDispatch();
    return (
        <>
            <Paper elevation={4} square>
                <Grid container justify="center">
                    <Grid item xs={6}>
                        <img className="MainLogo" alt="logo" src={logo} />
                    </Grid>
                    <Grid
                        item
                        xs={5}
                        container
                        justify="flex-end"
                        alignItems="center"
                    >
                        <Button component={Link} to="/map">
                            Карта
                        </Button>
                        <Button component={Link} to="/profile">
                            Профиль
                        </Button>
                        <Button
                            onClick={() => {
                                dispatch(clearRoute());
                                dispatch(loggedOut());
                            }}
                        >
                            Выйти
                        </Button>
                    </Grid>
                </Grid>
            </Paper>
        </>
    );
};

export default Header;
