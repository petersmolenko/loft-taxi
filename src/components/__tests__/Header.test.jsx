import React from "react";
import { fireEvent, render } from "@testing-library/react";
import renderer from "react-test-renderer";
import Header from "../Header";
import { Router } from "react-router-dom";
import configureStore from "redux-mock-store";
import { Provider } from "react-redux";
import { createBrowserHistory } from "history";
import { loggedOut } from "../../redux/modules/auth";

const mockStore = configureStore([]);

describe("Header testing", () => {
    it("renders correctly", () => {
        const store = mockStore({
            auth: { isLoggedIn: true },
        });
        store.dispatch = jest.fn();

        const history = createBrowserHistory();
        const tree = renderer
            .create(
                <Provider store={store}>
                    <Router history={history}>
                        <Header />
                    </Router>
                </Provider>
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("logout", () => {
        const store = mockStore({
            auth: { isLoggedIn: true },
        });
        store.dispatch = jest.fn();

        const history = createBrowserHistory();
        const { getByText } = render(
            <Provider store={store}>
                <Router history={history}>
                    <Header />
                </Router>
            </Provider>
        );

        fireEvent.click(getByText("Выйти"));
        expect(store.dispatch).toHaveBeenCalledTimes(1);
        expect(store.dispatch).toHaveBeenCalledWith(loggedOut());
    });

    it("map button click", () => {
        const store = mockStore({
            auth: { isLoggedIn: true },
        });
        store.dispatch = jest.fn();

        const history = createBrowserHistory();
        const { getByText } = render(
            <Provider store={store}>
                <Router history={history}>
                    <Header />
                </Router>
            </Provider>
        );

        fireEvent.click(getByText("Карта"));
        console.log(history.location.pathname);
        expect(history.location.pathname).toBe("/map");
    });
    it("profile button click", () => {
        const store = mockStore({
            auth: { isLoggedIn: true },
        });
        store.dispatch = jest.fn();

        const history = createBrowserHistory();
        const { getByText } = render(
            <Provider store={store}>
                <Router history={history}>
                    <Header />
                </Router>
            </Provider>
        );

        fireEvent.click(getByText("Профиль"));
        console.log(history.location.pathname);
        expect(history.location.pathname).toBe("/profile");
    });
});
