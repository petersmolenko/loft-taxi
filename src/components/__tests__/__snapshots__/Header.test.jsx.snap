// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Component Происходит корректная отрисовка 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation4"
>
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-justify-xs-center"
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
    >
      <img
        alt="logo"
        className="MainLogo"
        src="logo.svg"
      />
    </div>
    <div
      className="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-align-items-xs-center MuiGrid-justify-xs-flex-end MuiGrid-grid-xs-5"
    >
      <a
        aria-disabled={false}
        className="MuiButtonBase-root MuiButton-root MuiButton-text"
        href="/map"
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        role="button"
        tabIndex={0}
      >
        <span
          className="MuiButton-label"
        >
          Карта
        </span>
      </a>
      <a
        aria-disabled={false}
        className="MuiButtonBase-root MuiButton-root MuiButton-text"
        href="/profile"
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        role="button"
        tabIndex={0}
      >
        <span
          className="MuiButton-label"
        >
          Профиль
        </span>
      </a>
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-text"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          Выйти
        </span>
      </button>
    </div>
  </div>
</div>
`;
